### PR TITLE
Order quantity promotion rule

### PR DESCRIPTION
**Pull request by AlessioRocco**
This pull request add a promotion rule to allow promotion only for orders where quantity matches the given criteria. Almost the same of `ItemTotal` rule but based on order quantity.

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://alessiorocco-master-order-quantity-promotion.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://alessiorocco-master-order-quantity-promotion.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: core/spec/models/spree/promotion/rules/item_quantity_spec.rb [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+          expect(rule.eligibility_errors.full_messages.first).
+            to eq "This coupon code can't be applied to orders with less than 4 items."
+        end
+      end
+    end
+
+    context "and quantity is equal to the prefered maximum quantity" do
+      before { allow(order).to receive_messages quantity: 10 }
+
+      it "should not be eligible" do
```
Comment (JDutil) [inline core/spec/models/spree/promotion/rules/item_quantity_spec.rb]: should be*

File: core/spec/models/spree/promotion/rules/item_quantity_spec.rb [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+          allow(order).to receive_messages quantity: 6
+          expect(rule).to be_eligible(order)
+        end
+      end
+
+      context "and quantity is equal to the prefered minimum quantity" do
+
+        before { allow(order).to receive_messages quantity: 4 }
+
+        it "should not be eligible" do
```
Comment (JDutil) [inline core/spec/models/spree/promotion/rules/item_quantity_spec.rb]: should be*

File: core/spec/models/spree/promotion/rules/item_quantity_spec.rb [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
```
Comment (JDutil) [inline core/spec/models/spree/promotion/rules/item_quantity_spec.rb]: should be*

File: core/spec/models/spree/promotion/rules/item_quantity_spec.rb [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+          expect(rule.eligibility_errors.full_messages.first).
+            to eq "This coupon code can't be applied to orders with items quantity less than or equal to 4."
+        end
+      end
+    end
+
+    context "and quantity is equal to the prefered maximum quantity" do
+      before { allow(order).to receive_messages quantity: 10 }
+
+      it "should not be eligible" do
```
Comment (JDutil) [inline core/spec/models/spree/promotion/rules/item_quantity_spec.rb]: should be*

**Comment by gmacdougall** [comment]
Thanks for your work on this.

Does anyone have thoughts in to whether this is a desired feature for within Spree core? It's not something that any of the clients I've worked with has ever asked about.

**Comment by AlessioRocco** [comment]
I’m currently working on a project for a client that needs a Promotion that applies a X% off to lesser expensive product if a user buys X quantity. I did a pull request because I think that this rule mixed with others can be useful for different kind of promotion.

**Comment by JDutil** [comment]
I see the utility in it, but I don't imagine this is very popular either.  Generally I see promotions where the quantity is based on the quantity of a particular item not just the quantity of all items in your order combined.

**Comment by AlessioRocco** [comment]
@JDutil I think you can use this rule mixed with the [Product rule](https://github.com/spree/spree/blob/master/core/app/models/spree/promotion/rules/product.rb) to make the promotion that you suggested. Personally I prefer minimal `rules` and `actions` that a store Administrator can combine to create custom promotions instead to have big `rules` and `actions` that can be used just for some kind of promotions and need a developer to be created.

File: core/app/models/spree/promotion/rules/item_quantity.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        OPERATORS_MAX = ['lt','lte']
+
+        def applicable?(promotable)
+          promotable.is_a?(Spree::Order)
+        end
+
+        def eligible?(order, options = {})
+          quantity = order.quantity
+
+          lower_limit_condition = quantity.send(preferred_operator_min == 'gte' ? :>= : :>, preferred_quantity_min)
```
Comment (BenMorganIO) [inline core/app/models/spree/promotion/rules/item_quantity.rb]: That is a really cool way of using `#send`.

**Comment by JDutil** [comment]
That wouldn't work because this is based upon the order quantity not a line item quantity.

**Comment by AlessioRocco** [comment]
That's right, it was just an example to try to better explain the second point and that this rule, like others, can be used in different ways. 
Please, just let me understand what kind of rules you would like to add to spree core, I'm currently working on a lot on promotions for a client and I could do some pull requests for rules and actions but I would need to understand the spree way, what is preferred to be in the core and what is better to put into separate extensions. Thanks

**Comment by JDutil** [comment]
I agree on it being good to have lots of simple rules, actions, and calculators.  I'm interested in adding more if people find them useful, but there is the overhead of maintaining them so want to make sure people are interested in them first.

If you're doing a lot of promotions you may be interested in contributing specs for these calculators that you may find useful:
https://github.com/spree/spree/pull/4516
https://github.com/spree/spree/pull/4581

**Comment by peterberkenbosch** [comment]
Personally I would like this to be a spree-contrib repository where we can collect and manage them. Keeping the core lightweight but still provide the extended promotions users can use from the spree-contrib.

**Comment by BenMorganIO** [comment]
@peterberkenbosch I'm thinking promotions inside of spree could be managed by a `spree_promotions` gem.

**Comment by peterberkenbosch** [comment]
yes, something like that indeed.

**Comment by AlessioRocco** [comment]
:+1: for the `spree_promotions` gem.

**Comment by BenMorganIO** [comment]
@peterberkenbosch @AlessioRocco What I would really enjoy about the gem is that its goal would be simple to gather and become a library. Accepting promotions that ~10% of developers would use such as the spree_gateway gem.

**Comment by JDutil** [comment]
If someone would like to start one I can create a repo on https://github.com/spree-contrib/

**Comment by BenMorganIO** [comment]
@JDutil I was hoping it could be officially supported and move Spree's current promotions to that library or at least keep only two or three for core.

**Comment by AlessioRocco** [comment]
@JDutil ok, when you'll create the repo I’ll move this PR from here to the new gem. Thanks.

**Comment by JDutil** [comment]
We can't officially support everything.  If we're going to support something it's easier to maintain in core than it would be to open up a new repository to contain dozens of additional promotional rules/actions in a looser manner.

**Comment by wuboy0307** [comment]
+1 for spree_promotion gem.

I create many promotion actions and rules by my own. I would love to contribute them!

**Comment by bbonislawski** [comment]
I think that this is outdated already. If you think that it's not outdated, please mention me in comment and I'll reopen it

**Pull request closed without merging**